Adding contentprovider documentation (#692)
jupyter/repo2docker: docs/source/architecture.md

### Push

Optionally, repo2docker can **push** a built image to a [docker registry](https://docs.docker.com/registry/).
This is done as a convenience only (since you can do the same with a `docker push` after using repo2docker
only to build), and implemented in `Repo2Docker.push` method. It is only activated if using the
`--push` commandline flag.

### Run

Optionally, repo2docker can **run** the built image and allow the user to access the Jupyter Notebook
running inside by default. This is also done as a convenience only (since you can do the same with `docker run`
after using repo2docker only to build), and implemented in `Repo2Docker.run`. It is activated by default
unless the `--no-run` commandline flag is passed.

## ContentProviders

ContentProviders provide a way for `repo2docker` to know how to find and
retrieve a repository. They follow a similar pattern as the BuildPacks
described above. When `repo2docker` is called, its main argument will be
a path to a repository. This might be a local path or a URL. Upon being called,
`repo2docker` will loop through all ContentProviders and perform the following
commands:

> * Run the `detect(` method on the repository path given to `repo2docker`. This

Reviewer: ```suggestion
* Run the `detect()` method on the repository path given to `repo2docker`. This
```
I've not seen the `detect(` notation before and find it a bit weird, it feels like writing a sentence with no punctuation at the end 

Somehow unfinished :)

>   should return `True` if the path matches what the ContentProvider is looking

Reviewer: it should return something that `is not None`. So for example you could return an empty dictionary which is `False` (try `bool({})`) but `{} is not None == True`. The return value is then used as the argument to `fetch`. This allows detect to modify/change/clean up/add to the argument passed in from the CLI.

>   should return `True` if the path matches what the ContentProvider is looking

Author: ```suggestion
  should return any value other than `None` if the path matches what the ContentProvider is looking
```

  for.

  > For example, the [`Local` ContentProvider](https://github.com/jupyter/repo2docker/blob/80b979f8580ddef184d2ba7d354e7a833cfa38a4/repo2docker/contentproviders/base.py#L64)
  > checks whether the argument is a valid local path. If so, then `detect(`

>   > returns true.

Author: ```suggestion
  > returns True.
```

is this correct?

Reviewer: No. All content providers should (and do) return `None` if they don't match and something "that isn't `None`" when they do match. What precisely it is they return is left to them. The only requirement is that the `fetch()` method of the provider knows what to do with it.

Author: I was asking in this case specifically about the `Local` contentprovider rather than in general. I'll dig into it a bit more to figure it out

Author: see updated edit above, LMK if that's OK

Reviewer: > was asking in this case specifically

Ah ok. I missed that. 

Yes the suggestion is correct.

Reviewer: Do you have any idea why I can't commit the suggestion? Normally it works but it seems for this PR it consistently doesn't (and there are other PRs where it doesn't). So confusing.

> * If `detect(` returns `True`, run `fetch(` with the input path. This should

Author: ```suggestion
* If `detect()` returns `True`, run `fetch(` with the input path. This should
```

Reviewer: Apparently I can't commit these :( but I would if I could :)

<img width="664" alt="Screen Shot 2019-05-28 at 10 17 57" src="https://user-images.githubusercontent.com/1448859/58462329-ecef5800-8131-11e9-9e68-fbccd46b2938.png">
